hw/drivers/sensors/lis2dw12/src/lis2dw12.c

```diff
@@ -1946,89 +1943,106 @@ init_intpin(struct lis2dw12 * lis2dw12, hal_gpio_irq_handler_t handler,
     if (rc != 0) {
         LIS2DW12_ERR("Failed to initialise interrupt pin %d\n", pin);
         return rc;
-    } 
+    }
 
     return 0;
 }
 
 static int
-enable_interrupt(struct sensor * sensor, uint8_t int_to_enable)
+enable_interrupt(struct sensor *sensor, uint8_t int_to_enable, uint8_t int_num)
 {
     struct lis2dw12 *lis2dw12;
-    struct lis2dw12_private_driver_data *pdd;
+    struct lis2dw12_pdd *pdd;
     struct sensor_itf *itf;
     uint8_t reg;
     int rc;
 
+    if (!int_to_enable) {
+        rc = SYS_EINVAL;
+        goto err;
+    }
+
     lis2dw12 = (struct lis2dw12 *)SENSOR_GET_DEVICE(sensor);
     itf = SENSOR_GET_ITF(sensor);
     pdd = &lis2dw12->pdd;
 
     rc = lis2dw12_clear_int(itf, &reg);
     if (rc) {
-        return rc;
+        goto err;
     }
-    
+
     /* if no interrupts are currently in use enable int pin */
-    if (pdd->int_enable == 0) {
-        hal_gpio_irq_enable(itf->si_ints[pdd->int_num].host_pin);
+    if (!pdd->int_enable) {
+        hal_gpio_irq_enable(itf->si_ints[int_num].host_pin);
 
         rc = lis2dw12_set_int_enable(itf, 1);
         if (rc) {
-            return rc;
+            goto err;
         }
     }
 
-    /*update which interrupts are enabled */
-    pdd->int_enable |= int_to_enable;
-    
+    pdd->int_enable++;
+
     /* enable interrupt in device */
-    rc = lis2dw12_set_int1_pin_cfg(itf, pdd->int_enable);
+    if (int_num == 0) {
+        rc = lis2dw12_set_int1_pin_cfg(itf, int_to_enable);
```

Comment:
Doesn't lis2dw12_set_int1_pin_cfg( ) need to read the register contents and "OR" it with the next interrupt? It doesn't right now and currently wipes the register clean it is invoked.

lis2dw12_set_int1_pin_cfg( ) is called in two different places:

- lis2dw12_config ( ) 
- lis2dw12_sensor_set_notification ( ) which invokes enable_interrupt( ) which then invokes lis2dw12_set_int1_pin_cfg ()

Thoughts?

Reply:
I see what you mean here, `lis2dw12_set_int1_pin_cfg()` probably should just OR it then if we were to support multiple APIs from enabling interrupts and we can add a interrupt clear API to clear the int cfg. The initial thought was, notification/config would just override each other but that sounds like might be too limiting.

Reply:
Previously it was ORed in the enable_interrupt function. If it should be done in the set_int1_pin_cfg function that is also fine. I believe it is quite important because without the ORing if a stream read is performed when even a single notification is registered the stream read will reset the interrupt for the notification.